### ISSUE TITLE
[XGrid] Add support for interactive 'filter execution mode'

### ISSUE THREAD

**Issue by sebastianfrey** [mui-org/material-ui-x#1723]
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Summary 💡

It would be nice to have the ability to optionally control whether a filter is applied immediatly or through an additional user interaction like a button click. I will name this feature `filter execution mode`

Therefore a new grid property `filterExecutionMode` is propsed. This property is optional an accepts two values:

1. `immediate` - represents the current behaviour and is the default mode how filters are executed
2. `interactive` - requires an additional user interaction to execute filters

When `filterExecutionMode` is set to `interactive` an additional Button `Apply` is rendered in the filter panel. Then in order to execute a given filter, a user must click this button.

## Examples 🌈

Execute filters immediatly:

```jsx
<XGrid
   ...
   filterExecutionMode="immediate"
/>
```
Execute filters interactive:

```jsx
<XGrid
   ...
   filterExecutionMode="interactive"
/>
```

## Motivation 🔦

When you do server side querying an immediate filter execution is not always the best solution, especially when

- the data set to query is very large
- a user creates a complex filter condition
- you must avoid redundant server calls

The proposed filter exeution mode would solve those previously named scenarios.

## Order id 💳

22075

**Comment by oliviertassinari** [comment]
This is an interesting concern. From what I understand, you have to set `filterMode="server"` to perform the filtering server side when the `onFilterModelChange` event fire.
Then, you could customize the filter panel to add a confirmation button. And finally, you will need to add logic to make the UX is intuitive, for instance, to make it clear if the filtering is applied or not.

**Comment by sebastianfrey** [comment]
That pretty sums it up. In addition I would propose, that this feature might be independent from the `filterMode`. 

The `autoOk` and `autoSelect`approach is far more intuitive than the proposed `fitlerExecutionMode`. So when one wants his users to submit filter changes, you would write `autoSubmitFilter={false}`.

**Comment by nalbhike** [comment]
Very important and basic feature. 
This seems very straightforward and easy to implement, any chance we can move forward with this?

**Comment by sebastianfrey** [comment]
Several different customers have recently requested this feature from us.  How can we move forward here?

@oliviertassinari  Are you willing to accept a community PR for this feature? Or do you think that goes too deep and should only be handled by one of you guys?

**Comment by sebastianfrey** [comment]
Hi @cherniavskii, I am reaching out to you, because it on the MUI-X Website  you are marked as responsible tech lead for the data grid. Maybe you can help here regarding my comment two weeks ago?

**Comment by cherniavskii** [comment]
Hey @sebastianfrey 
Community PR is definitely welcome, we can provide with guidance if necessary.

Regarding the implementation direction, my proposal would be to store the "draft" filter model outside of the data grid state.
It'll be up to developers to manage the draft filter model and sync the active filter model with it (e.g. on user click).

To make this possible, we need to expose a component like `GridFilterPanel`, but that component should not rely on the data grid state, and accept filter model value/onChange props. It should be extracted from `GridFilterPanel`, with no breaking changes to the `GridFilterPanel` itself.
The main goal of this new component is to allow devs reuse existing filter panel without rebuilding it, while being able to wire it up to a different source of the state.

What do you think?

**Comment by sebastianfrey** [comment]
Hi @cherniavskii,

I looked a bit into `GridFilterPanel` and its direct descendants like ` GridFilterForm` and or some of the `GridFilterInput*` components.

It seems that these descendants components are already isolated from the internal grid filter state. Grid filter state related stuff is handled by `GridFilterPanel` itself. Only column information or translations are fetched through `apiRef` by those components. This seems to be fine as long as the customer filter panel is rendered under the grids context. If a consumer wants to render the custom filter panel component outside of the grids context, the consumer must be able to pass down an `apiRef` instance.

Overall its seems pretty doable to derive a new component from current `GridFilterPanel` state, which can be used internally and externally.

What do you think about how the `apiRef` is used in descendants components currently? Is it enough to isolate grid filter state related access, or should the column information and translation related `apiRef` access also be targeted?

**Comment by cherniavskii** [comment]
> Is it enough to isolate grid filter state related access, or should the column information and translation related apiRef access also be targeted?

Do you have a use case in mind, where taking column information from the data grid context wouldn't work?
I can't think of one TBH, so I'd say isolating only filter state related stuff should be enough.

**Comment by sebastianfrey** [comment]
No, I also don't have a use case in mind. Just wanted to clarify this, before I start implementing a solution.

**Comment by cherniavskii** [comment]
Alright, sounds like a plan then 👍🏻
Let me know if you need any help.

**Comment by sebastianfrey** [comment]
@cherniavskii after diving deeper I have noticed, that `GridFilterForm` selects the grid filter state from apiRef. So if I introduce a `GridFilterPanelBase` component with `value/onChange` props as discussed, in a controlled state scenario, the grid filter state access in `GridFilterForm` might be stale.

To workaround this,  I thought about introducing a new `filterModel` prop on `GridFilterForm`. Here I am unsure, how you want this prop to behave. Do you want this new prop to be required or optional? 

1. If I make it required this is a breaking change for `GridlFilterForm`, but it might be cleaner.
2. If I make it optional, the effective grid filter state has to be derived conditionally in the `GridFilterForm`, which is less clean and might cause unexpected behavior, when not used properly.

What do you think? Which way would you go here?

**Comment by cherniavskii** [comment]
@sebastianfrey How about extracting `GridFilterFormBase` that receives filtering state via props?

**Comment by sebastianfrey** [comment]
@cherniavskii simple and elegant. Would you make `GridFilterFormBase` public or privat? Right now I don't see a use case where it might be useful to have `GridFilterFormBase` public available. What do you think?

**Comment by cherniavskii** [comment]
> Right now I don't see a use case where it might be useful to have GridFilterFormBase public available

Then definitely private 👍🏻
Would you also like to add a recipe to https://mui.com/x/react-data-grid/filtering-recipes/ showing off the filter execution behavior?

**Comment by sebastianfrey** [comment]
Yes, I will add a new sample, show casing the controlled filter panel approach.

**Comment by sebastianfrey** [comment]
@cherniavskii while crafting the sample, I stumbled about the following thing:

My current implementation of `GridFilterPanelBase` owns the footer section. This replicates the current behavior of `GridFilterPanel`. So if a consumer want's to have a separate footer section, he would build a component like this:

```tsx
<Box sx={{ display: 'flex', flexDirection: 'column' }}>
  <GridFilterPanelBase
    {...props}
    filterModel={draftFilterModel}
    onFilterModelChange={setDraftFilterModel}
  />
  <CustomFooter />
</Box>
```

But this will result in a semantically incomplete component tree. If you expand `GridFilterPanelBase`'s tree mentally, you get this:

```tsx
<Box sx={{ display: 'flex', flexDirection: 'column' }}>
  <GridPanelWrapper>
    <GridPanelContent>
       ...
    <GridPanelContent>
  </GridPanelWrapper>
  <CustomFooter />
</Box>
```

So the `GridPanelWrapper` is wrapped by a "wrapper".  In order to support consumers to provide a custom footer section, and keeping the semantic layout structure intact at the same time, I can imagine two approaches:

1. Introducing a slot named `footer` in `GridFilterPanelBase`. This slot can be leveraged to pass a custom footer section to `GridFilterPanelBase`.
2.  Make `GridFilterPanel` more composable, and dissect it further into smaller base components.  IMO this is more risky and might introduce new bugs.

Any thoughts on this? Are you fine with my first approach? Or should I head into a different direction? Or do you think this concern is negligible?

------------

Edit: It works totally fine right now. So I think we should not make it more complicated than needed.
